src/core/unit.c

```diff
@@ -1575,6 +1605,12 @@ int unit_start(Unit *u) {
         if (!unit_supported(u))
                 return -EOPNOTSUPP;
 
+        /* Let's make sure that the deps really are in order before we start this. Normally the job engine should have
+         * taken care of this already, but let's check this here again. After all, the dependencies might have changed
+         * due to a reload or because the user forgot to place ordering dependencies in conjunction with BindsTo=. */
+        if (!unit_verify_deps(u))
+                return -ENOLINK;
```

Comment:
I'm reading the docs
> BindsTo=
 Configures requirement dependencies, very similar in style to Requires=, however in addition to this behavior, it also declares that this unit is stopped when any of the units listed
           suddenly disappears. Units can suddenly, unexpectedly disappear if a service terminates on its own choice, a device is unplugged or a mount point unmounted without involvement of systemd.

It's not so obvious why `BindsTo=` requires `After=`

Also, this commit changes the semantic of the `BindsTo=`:
```
-bash-4.3# systemctl cat a --no-pager
# /etc/systemd/system/a.service
[Service]
Type=oneshot
ExecStart=/bin/sh -x -c 'echo Hey'
RemainAfterExit=true

-bash-4.3# systemctl cat b --no-pager
# /etc/systemd/system/b.service
[Unit]
BindsTo=a.service

[Service]
ExecStart=/bin/sh -x -c 'echo B'
Type=oneshot
RemainAfterExit=true
```
`v229` works

this PR:
```
-bash-4.3# systemctl start b
A dependency job for b.service failed. See 'journalctl -xe' for details.

-bash-4.3# journalctl -u a -u b --no-pager
-- Logs begin at Thu 2016-11-24 22:05:00 UTC, end at Thu 2016-11-24 22:12:25 UTC. --
Nov 24 22:06:48 systemd-testsuite systemd[1]: b.service: Trying to enqueue job b.service/start/replace
Nov 24 22:06:48 systemd-testsuite systemd[1]: a.service: Installed new job a.service/start as 244
Nov 24 22:06:48 systemd-testsuite systemd[1]: b.service: Installed new job b.service/start as 194
Nov 24 22:06:48 systemd-testsuite systemd[1]: b.service: Enqueued job b.service/start as 194
Nov 24 22:06:48 systemd-testsuite systemd[1]: b.service: Job b.service/start finished, result=dependency
Nov 24 22:06:48 systemd-testsuite systemd[1]: Dependency failed for b.service.
Nov 24 22:06:48 systemd-testsuite systemd[1]: b.service: Job b.service/start failed with result 'dependency'.
Nov 24 22:06:48 systemd-testsuite systemd[1]: a.service: Passing 0 fds to service
Nov 24 22:06:48 systemd-testsuite systemd[1]: a.service: About to execute: /bin/sh -x -c 'echo Hey'
Nov 24 22:06:48 systemd-testsuite systemd[1]: a.service: Forked /bin/sh as 227
Nov 24 22:06:48 systemd-testsuite systemd[1]: a.service: Changed dead -> start
Nov 24 22:06:48 systemd-testsuite systemd[227]: a.service: Executing: /bin/sh -x -c 'echo Hey'
Nov 24 22:06:48 systemd-testsuite systemd[1]: Starting a.service...
Nov 24 22:06:48 systemd-testsuite systemd[1]: b.service: Collecting.
Nov 24 22:06:48 systemd-testsuite sh[227]: + echo Hey
Nov 24 22:06:48 systemd-testsuite sh[227]: Hey
Nov 24 22:06:48 systemd-testsuite systemd[1]: a.service: Child 227 belongs to a.service
Nov 24 22:06:48 systemd-testsuite systemd[1]: a.service: Main process exited, code=exited, status=0/SUCCESS
Nov 24 22:06:48 systemd-testsuite systemd[1]: a.service: Changed start -> exited
Nov 24 22:06:48 systemd-testsuite systemd[1]: a.service: Job a.service/start finished, result=done
Nov 24 22:06:48 systemd-testsuite systemd[1]: Started a.service.
Nov 24 22:06:48 systemd-testsuite systemd[1]: a.service: cgroup is empty
Nov 24 22:09:35 systemd-testsuite systemd[1]: b.service: Collecting.
Nov 24 22:09:35 systemd-testsuite systemd[1]: b.service: Collecting.
```
And I'm not sure this is the right thing to do